File: javascript/grade-school.js
export class GradeSchool {
  constructor() {
    this.school = {};
  }
  roster() {
    let res = {};
    for (let [k, v] of Object.entries(this.school)) {
      res[k] = [...v].sort();
    }
    return res;
  }

  add(name, grade) {
    if (this.school[grade]) {
      this.school[grade].push(name);
    } else {
      this.school[grade] = [name];
    }
  }
  grade(g) {
    if (this.school[g]) {
      return [...this.school[g] ].sort();
    } else {
      return [];
    }
  }
}

File: F#/GradeSchool.fs
module GradeSchool

type School = Map<int, string list>

let empty : School = Map.empty

let addToOptionalList student =
    function
    | None -> Some [ student ]
    | Some list -> Some <| student :: list

let add (student: string) (grade: int) (school: School) : School =
    Map.change grade (addToOptionalList student) school


let roster (school: School) : string list =
    school
    |> Map.toList
    |> List.collect (snd >> List.sort)

let grade (number: int) (school: School) : string list =
    if Map.containsKey number school then
        Map.find number school
    else
        []
